synapse/handlers/e2e_keys.py

```diff
@@ -542,6 +546,50 @@ async def on_federation_query_client_keys(
 
         return ret
 
+    async def claim_one_time_keys_local(
+        self, local_query: List[Tuple[str, str, str]]
+    ) -> Iterable[Dict[str, Dict[str, Dict[str, JsonDict]]]]:
+        """Claim one time keys for local users.
+
+        1. Attempt to claim OTKs from the database.
+        2. Ask application services if they provide OTKs.
+        3. Attempt to fetch fallback keys from the database.
+
+        Args:
+            query_list: An iterable of tuples of (user ID, device ID, algorithm).
+
+        Returns:
+            An iterable of maps of user ID -> a map device ID -> a map of key ID -> JSON bytes.
+        """
+        otk_results, not_found = await self.store.claim_e2e_one_time_keys(local_query)
+
+        # If the application services have not provided any keys via the C-S
+        # API, query it directly.
+        if self._query_appservices_for_otks:
+            # Partition the list into things an appservice might even care about.
+            as_query = []
+            non_as_query = []
+            for user_id, device, algorithm in not_found:
+                if self.store.get_if_app_services_interested_in_user(user_id):
+                    as_query.append((user_id, device, algorithm))
+                else:
+                    non_as_query.append((user_id, device, algorithm))
+
+            (
+                appservice_results,
+                not_found,
+            ) = await self.appservice_handler.claim_e2e_one_time_keys(as_query)
+
+            # TODO This probably horribly inefficient.
+            not_found.extend(non_as_query)
```

Comment:
I don't really know where to put this comment, but a potential consideration is the appservice returning *less* keys than the request: in this case, the homeserver should use the fallback key (I'll add this to the MSC).

For example, a request for:
```json
{
  "@alice:example.org": {
    "DEVICEID": ["signed_curve25519", "signed_curve25519"]
  },
  "@bob:example.org": {
    "DEVICEID": ["signed_curve25519"]
  }
}
```

might be returned with: 

```jsonc
{
  "@alice:example.org": {
    "DEVICEID": {
      "signed_curve25519:AAAAHg": { ... }
    }
  }
}
```

The second key for Alice and Bob's only key should both be the respective fallback keys. 

I'm also unsure if Synapse could even generate such a situation, though.

Reply:
Makes sense, the API allows this, I'll just need to implement diffing the returned values from the query.

Reply:
This should be fixed (with a test!) 👍